File: remote/lua/robot.lua
local socket = require("socket")

robot = {}

function robot.open_remote(host, port)

	local client, err = socket.connect(host, port)

	if client == nil then
		print("failed to connect to remote: " .. err)
		return false	
	end

	robot.client = client

	return true
end

function robot.close_remote()

	if robot.client == nil then
		print("not connected to remote")
		return false	
	end

	robot.client:close()
	
	return true
end

function robot.drive(speed_a, speed_b)
	
	local success = false

	if robot.client == nil then
		print("not connected to remote")
		return false	
	end

	local a = string.format("%+04d", speed_a)
	local b = string.format("%+04d", speed_b)

	local cnt, err = robot.client:send("d" .. a .. b .."\n")

	if cnt == nil then
		print("sending to remote failed: " .. err)
	else
		local line, err = robot.client:receive()

		if line == nil then
			print("receiving from remote failed: " .. err)
		elseif line == "OK" then
			success = true	
		end
	end

	return success 
end

function robot.drive_forward(speed)

	speed = speed or 100

	return robot.drive(speed, speed)
end

function robot.drive_backward(speed)

	speed = speed or 100

	return robot.drive(-speed, -speed)
end

function robot.turn_left(speed)

	speed = speed or 100

	return robot.drive(-speed, speed)
end

function robot.turn_right(speed)

	speed = speed or 100

	return robot.drive(speed, -speed)
end

function robot.stop()

	return robot.drive(0, 0)
end

function robot.sees_obstacle()

	if robot.dummy_cnt == nil then
		robot.dummy_cnt = 0
	else
		robot.dummy_cnt = robot.dummy_cnt + 1
	end

	if robot.dummy_cnt > 10 and robot.dummy_cnt < 20 then
		print("BOOOM!")
		return true
	elseif robot.dummy_cnt == 20 then
		robot.dummy_cnt = 0
	end

	return false
end

local clock = os.clock

function sleep(n)  -- seconds
   local t = clock()
   while clock() - t <= n do end
end

File: remote/python/robot.py
import telnetlib

__author__ = 'stefan'

# TODO: error handling (response from robot is "OK" or "ERROR" => map to True or False


class Robot:

    def __init__(self, host="ESP_06F38E"):
        self.tn = telnetlib.Telnet(host)

    def drive(self, speed_a, speed_b):
        self.tn.write("d%+04d%+04d\n" % (speed_a, speed_b))
        id, re, txt = self.tn.expect([r"OK\n", r"ERROR\n"])
        return id == 0

    def drive_forward(self, speed=100):
        return self.drive(speed, speed)

    def drive_backward(self, speed=100):
        return self.drive(-speed, -speed)

    def turn_left(self, speed=100):
        return self.drive(-speed, speed)

    def turn_right(self, speed=100):
        return self.drive(speed, -speed)

    def stop(self, speed=0):
        return self.drive(0, 0)

    def connect_scratch(self, host_ip):
        self.tn.write("Sc%s" % host_ip)
        id, re, txt = self.tn.expect([r"OK\n", r"ERROR\n"])
        return id == 0

    def disconnect_scratch(self):
        self.tn.write("Sd")
        id, re, txt = self.tn.expect([r"OK\n", r"ERROR\n"])
        return id == 0

    def dist(self):
        self.tn.write("r")
        id, re, txt = self.tn.expect([r"ERROR\n", r"[-+]?[0-9]*\.?[0-9]+\n"])

        if id != 0:
           return float(txt)

        return 0.0 

    def sees_obstacle(self):
        self.tn.write("o")
        id, re, txt = self.tn.expect([r"NO\n", r"FAR\n", r"CLOSE\n"])
        return id
